### PR TITLE
Fixes part of #2423: Solve naming convention for whole number type

### LINKED ISSUE #2423: Fix Mathematical Naming in classify rules
**Describe the bug**
Some of the variables naming related to numbers in `src/main/java/org/oppia/android/domain/classify/rules` and it's test packet are mathematical incorrect. 
example - `NEGATIVE_WHOLE_NUMBER_VALUE_TEST_2` where a whole number cannot be negative. 

**Expected behavior**
Need to fix all such variables name

Follow these comments to solve this issue - https://github.com/oppia/oppia-android/pull/2443#issuecomment-758299254
Reach out to @anandwana001 before start working on this. 

### How to work on this issue
Suggesting few points below on how to solve this particular issue:

1. Let's have a doc where you can list all the types of the number we have.
2. Specify the rules based on those numbers. Example - A whole number can't be negative.
3. You can add rules for all the types of number like a fraction, mixed fraction, Ratio of numbers etc...
4. After completing with doc, either we can fix everything in this PR, but I guess it might make this PR big, so we can divide it into multiple sub-issues.

### PR DESCRIPTION
**Pull request by codersanjeev**
## Explanation

Fixes part of #2423. Solve naming convention for whole number type. Fix whole number function call at various places by removing isNegative parameter from the function call.

Doc Link: https://docs.google.com/document/d/1ld3wdeS61-_fery79mcCXczdSYJzSFeyi7360DlIR1c/edit?usp=sharing
## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by Luffy18346** [comment]
@codersanjeev nit change
PR title should start with "Fix #2423" instead of "Fixes #2423" and after correcting the title, please check 1st checkbox in the checklist.

**Comment by anandwana001** [comment]
@codersanjeev We also need to check all other places where we are mentioning wrong names and fixed to Real Number/Fraction/Mixed Numbers/Rational Numbers. 
For this, you need to check the whole InputInteraction part.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 1]
```diff
@@ -125,7 +125,7 @@ class FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest {
 
   @Test
   fun testEquivalentAndSimplest_wholeNumber2Answer_negativeWholeNumber2Input_doesNotMatch() {
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: Please also update the test name to match the input.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @codersanjeev!

Are there any other variable names in the codebase that need to correspondingly be updated? I think #2423 is tracking all of the classifier files that incorrectly reference types of numbers. Are there others?

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt (hunk 1 above)
Comment (codersanjeev) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: Thanks for the feedback, yes, I need to check for some other variable names also and i will update the test name to match the input.

**Comment by anandwana001** [comment]
once the PR update with other parts of name fixing as per Ben's comment above, assign me back. I will review it.

**Comment by codersanjeev** [comment]
> once the PR update with other parts of name fixing as per Ben's comment above, assign me back. I will review it.
I think createWholeNumber function in InteractionObjectTestBuilder should not have isNegative parameter as whole numbers are always zero or positive, we can either remove this parameter or make it false by default as it is being passed as false from all places except for NEGATIVE_INTEGER_VALUE_TEST_2.
what do you say, @anandwana001?

**Comment by chrk2205** [comment]
> > once the PR update with other parts of name fixing as per Ben's comment above, assign me back. I will review it.
> > I think createWholeNumber function in InteractionObjectTestBuilder should not have isNegative parameter as whole numbers are always zero or positive, we can either remove this parameter or make it false by default as it is being passed as false from all places except for NEGATIVE_INTEGER_VALUE_TEST_2.
> > what do you say, @anandwana001?

assign it to the reviewer so you can get the review @codersanjeev

**Comment by anandwana001** [comment]
@codersanjeev this sounds good. So, `createWholeNumber` is something where we should enforce to keep this negative else the number can't be a whole number.

**Comment by chrk2205** [comment]
> @codersanjeev this sounds good. So, `createWholeNumber` is something where we should enforce to keep this negative else the number can't be a whole number.

instead of  `createWholeNumber` we can refactor it to `createInteger`

**Comment by anandwana001** [comment]
Need to confirm first either we need whole number or integer!!

**Comment by codersanjeev** [comment]
> Need to confirm first either we need whole number or integer!!

I think we can't do that as we already have createInt at line number 121 in InteractionObjectTestBuilder class

**Comment by codersanjeev** [comment]
> @codersanjeev this sounds good. So, `createWholeNumber` is something where we should enforce to keep this negative else the number can't be a whole number.

I have pushed this change and updated the call to `createWholeNumber` function wherever required. Please review @anandwana001

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 2]
```diff
@@ -24,15 +24,13 @@ class FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest {
...
     )
   private val WHOLE_NUMBER_VALUE_TEST_2 =
     InteractionObjectTestBuilder.createWholeNumber(
-      isNegative = false,
       value = 2
     )
-  private val NEGATIVE_WHOLE_NUMBER_VALUE_TEST_2 =
+  private val NEGATIVE_INTEGER_VALUE_TEST_2 =
     InteractionObjectTestBuilder.createWholeNumber(
       isNegative = true,
```
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:35]: Could you help in understanding here why we are creating a negative whole number here?
As this is wrong and this is something we are solving in this issue.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
As for this PR, suggesting few points below on how to solve this particular issue:

1. Let's have a doc where you can list all the types of the number we have.
2. Specify the rules based on those numbers. Example - A whole number can't be negative. 
3. You can add rules for all the types of number like a fraction, mixed fraction, Ratio of numbers etc...
4. After completing with doc, either we can fix everything in this PR, but I guess it might make this PR big, so we can divide it into multiple sub-issues.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 3]
```diff
@@ -24,15 +24,13 @@ class FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest {
...
       value = 5
     )
   private val WHOLE_NUMBER_VALUE_TEST_2 =
     InteractionObjectTestBuilder.createWholeNumber(
-      isNegative = false,
       value = 2
     )
-  private val NEGATIVE_WHOLE_NUMBER_VALUE_TEST_2 =
+  private val NEGATIVE_INTEGER_VALUE_TEST_2 =
     InteractionObjectTestBuilder.createWholeNumber(
```
Comment (chrk2205) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: we are using `createWholeNumber` function to create a negative integer.
which is mathematically incorrect
@anandwana001 @coder2699

**Review by chrk2205 — CHANGES_REQUESTED** [review]

**Comment by Luffy18346** [comment]
@codersanjeev  Any updates on this?

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt (hunk 2 above)
Comment (codersanjeev) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:35]: I have updated this function call to createMixedNumber.

**Comment by codersanjeev** [comment]
> @codersanjeev Any updates on this?

yes, I have pushed the required code changes.

**Comment by anandwana001** [comment]
@coder2699 I suggest following the steps from the above comment. 
https://github.com/oppia/oppia-android/pull/2443#pullrequestreview-570913348

**Comment by codersanjeev** [comment]
> @coder2699 I suggest following the steps from the above comment.
> [#2443 (review)](https://github.com/oppia/oppia-android/pull/2443#pullrequestreview-570913348)

I am working on it.

**Comment by codersanjeev** [comment]
> As for this PR, suggesting few points below on how to solve this particular issue:
> 
> 1. Let's have a doc where you can list all the types of the number we have.
> 2. Specify the rules based on those numbers. Example - A whole number can't be negative.
> 3. You can add rules for all the types of number like a fraction, mixed fraction, Ratio of numbers etc...
> 4. After completing with doc, either we can fix everything in this PR, but I guess it might make this PR big, so we can divide it into multiple sub-issues.

I have started creating the doc [here](https://docs.google.com/document/d/1ld3wdeS61-_fery79mcCXczdSYJzSFeyi7360DlIR1c/edit?usp=sharing) as per your suggestions.
Please review @anandwana001

**Comment by anandwana001** [comment]
@codersanjeev please update the pr title and description based on which specific number type this pr is solving.

**Comment by Luffy18346** [comment]
@codersanjeev 
Please update **PR title** as "Fix #{bug num}: Solve naming convention for whole number type" 
and **PR Explanation** as "Fixes #{bug num}. Solve naming convention for whole number type. Fix whole number function call at various places by removing isNegative parameter from the function call."

**Comment by anandwana001** [comment]
@codersanjeev @Luffy18346 just an update here, as per our discussion over Google docs, this PR only focuses on one type of number type, as we have around 5-6 number type, each number type will be going to have a single PR as we need to fix single type at multiple places, at all the occurrences. 

So I suggest @codersanjeev to follow the checklist and update the PR title and description accordingly.

**Comment by codersanjeev** [comment]
> @codersanjeev @Luffy18346 just an update here, as per our discussion over Google docs, this PR only focuses on one type of number type, as we have around 5-6 number type, each number type will be going to have a single PR as we need to fix single type at multiple places, at all the occurrences.
> 
> So I suggest @codersanjeev to follow the checklist and update the PR title and description accordingly.

I have a doubt, in whole number-creation [here](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/domain/src/test/java/org/oppia/android/domain/classify/InteractionObjectTestBuilder.kt#L43) we should not be building it using a fraction as fraction part will always be (0/1) as written in the comment also, instead, shouldn't we use the function something like createInteger and passing a positive sign to createInteger function as whole number is a special type of integer with positive sign.

**Comment by Luffy18346** [comment]
> @codersanjeev @Luffy18346 just an update here, as per our discussion over Google docs, this PR only focuses on one type of number type, as we have around 5-6 number type, each number type will be going to have a single PR as we need to fix single type at multiple places, at all the occurrences.
> 
> So I suggest @codersanjeev to follow the checklist and update the PR title and description accordingly.

Ok @anandwana001 As per you described earlier there will be 5-6 PR for this single issue, which is why we are not adding "Fix #{bug num}", right?  but can we add "Fix part of #bugnum: ..." in Title and "Fixes part of #bugnum"  in Explanation?

**Comment by anandwana001** [comment]
@Luffy18346 Yes, we will be tracking the issue with the PR in parts.

**Comment by anandwana001** [comment]
> > @codersanjeev @Luffy18346 just an update here, as per our discussion over Google docs, this PR only focuses on one type of number type, as we have around 5-6 number type, each number type will be going to have a single PR as we need to fix single type at multiple places, at all the occurrences.
> > So I suggest @codersanjeev to follow the checklist and update the PR title and description accordingly.
> 
> I have a doubt, in whole number-creation [here](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/domain/src/test/java/org/oppia/android/domain/classify/InteractionObjectTestBuilder.kt#L43) we should not be building it using a fraction as fraction part will always be (0/1) as written in the comment also, instead, shouldn't we use the function something like createInteger and passing a positive sign to createInteger function as whole number is a special type of integer with positive sign.

Looks like a good catch @codersanjeev 
Could you please update the PR with the solution for this on how we can correct this?

**Comment by codersanjeev** [comment]
> > > @codersanjeev @Luffy18346 just an update here, as per our discussion over Google docs, this PR only focuses on one type of number type, as we have around 5-6 number type, each number type will be going to have a single PR as we need to fix single type at multiple places, at all the occurrences.
> > > So I suggest @codersanjeev to follow the checklist and update the PR title and description accordingly.
> > 
> > 
> > I have a doubt, in whole number-creation [here](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/domain/src/test/java/org/oppia/android/domain/classify/InteractionObjectTestBuilder.kt#L43) we should not be building it using a fraction as fraction part will always be (0/1) as written in the comment also, instead, shouldn't we use the function something like createInteger and passing a positive sign to createInteger function as whole number is a special type of integer with positive sign.
> 
> Looks like a good catch @codersanjeev
> Could you please update the PR with the solution for this on how we can correct this?

I have tried correcting the definition of createWholeNumber function as per observation, but this resulted in failing test cases of fractioninput which are using createWholeNumber function as it is expecting a fractional part in the whole number. I think we would need to re-write the failing test cases again as per new definition of createWholeNumber function?
One more question: How whole number is different from non negative integer which we are creating in createNonNegativeInt function. logically, only one of them should exist as both of them are representing same set of numbers.
@anandwana001 need your views on this.

**Comment by Luffy18346** [comment]
> @Luffy18346 Yes, we will be tracking the issue with the PR in parts.

@anandwana001 then based on this I think PR title and Explanation should be updated to 'Fixes part of #bugnum: ...'. WDYT?

**Comment by Luffy18346** [comment]
> > @Luffy18346 Yes, we will be tracking the issue with the PR in parts.
> 
> @anandwana001 then based on this I think PR title and Explanation should be updated to 'Fixes part of #bugnum: ...'. WDYT?

@anandwana001 WDYT?

**Comment by anandwana001** [comment]
Yes, absolutely. Fix part of, not fixes, as per the checklist.

**Comment by anandwana001** [comment]
@codersanjeev any update on this?

**Comment by codersanjeev** [comment]
> @codersanjeev any update on this?
https://github.com/oppia/oppia-android/pull/2443#issuecomment-772461563
stuck here

**Comment by anandwana001** [comment]
@codersanjeev
```
The whole numbers are the numbers 0, 1, 2, 3, 4, and so on (the natural numbers and zero). Negative numbers are not 
considered "whole numbers." All natural numbers are whole numbers, but not all whole numbers are natural numbers since
zero is a whole number but not a natural number.
```
`// Whole number fractions imply '0/1' fractional parts.`

Yes, we do need a lot of re-write here as the functions `createWholeNumber`, `createFractions,  etc need to go in a correct logical sense. 

From here, what I think we can do is, in the doc, we can create a table of all such incorrect functions and then list a requirement of what are the things we need to update to make them correct. Once the doc gets reviewed, we can jump to the implementation part. 

This issue is something we need to do in 2 steps:
1. Clear Documentation
2. Implementation

**Comment by codersanjeev** [comment]
got it, I will update the document. @anandwana001

**Comment by codersanjeev** [comment]
I have updated the doc @anandwana001, please review.

**Comment by anandwana001** [comment]
@codersanjeev could you please add the doc link in PR description?

**Comment by codersanjeev** [comment]
> @codersanjeev could you please add the doc link in PR description?

added

**Comment by anandwana001** [comment]
> From here, what I think we can do is, in the doc, we can create a table of all such incorrect functions and then list a requirement of what are the things we need to update to make them correct. Once the doc gets reviewed, we can jump to the implementation part.

> This issue is something we need to do in 2 steps:

> - Clear Documentation
> - Implementation

I guess we need more details in doc, about functions, what other functions are using them, what arguments passed and what approach you to suggest to fix them. Once we done with all this in the documentation, our implementation becomes easy.

**Comment by anandwana001** [comment]
@codersanjeev any update on this?

**Comment by rt4914** [comment]
Closing this because of no updates.

**Pull request closed without merging**